### ISSUE TITLE
EMU Danfysik: retrying setting setpoint takes too long [timebox: 2 days]

### ISSUE THREAD

**Issue by Tom-Willemsen** [ISISComputingGroup/IBEX#5838]
As an instrument scientist on EMU, I would like to eliminate long delays when setting a field.

EMU's danfysik has intermittent communications problems, where "about 10%" of the time it will fail to take a setpoint correctly. This is currently mitigated in the scripts using a retry loop. Using this retry loop, the danfysik does always *eventually* get to the correct field, but this can be delayed significantly by the retry logic (scientists think they lose about 10 minutes for each retry it has to do).

# Acceptance criteria
- The Danfysik reliably reaches the requested field with only short delays, even in the case where some setpoints are ignored

# Notes
- This might be fixed either in the danfysik ioc itself, or in EMU's `setmag` script.
- This will require testing on EMU. Testing time should be available during the short shutdown, if we can make the change in time

**Comment by kjwoodsISIS** [comment]
Is this a software problem or a hardware problem?
What happens if you directly set the setpoint from an EPICS console?

**Comment by Tom-Willemsen** [comment]
I believe it is a communication problem of some sort between EPICS and the danfysik. The PV is getting set, I believe the command is being sent down the serial line, but the danfysik is not acting on it (10% of the time). Whether this is because we have our comms settings wrong or the danfysik itself is unreliable, I don't know.

**Comment by John-Holt-Tessella** [comment]
I think if we had the communications wrong it would work either 100% or 0% unless it is a timing issue, e.g. send too many commands at once. Should we ask the electronics guys to check the cabels first?

**Comment by DominicOram** [comment]
Should have a look if this happens in the VI

**Comment by FreddieAkeroyd** [comment]
Also do we have a setpoint readback?

**Comment by DominicOram** [comment]
Is this just this one Danfysik or does it effect others?

**Comment by DominicOram** [comment]
Investigate a "correct" solution but use a script if it will take longer than the timebox

**Comment by JamesKingWork** [comment]
# Status so far

We have been running a script which sets fields in steps of 5 from 5 to 50 and then back down a number of times, and have had no setpoints not accepted and no retries are done. This is after 2 full runs of the script.

On Friday they had noticed the problem and there is something in the log file about protocol message mismatches, we are waiting on a reply from Peter about when the error occurred to see if it correlates. The error only seems to have occurred for a second before the matches passed again so a wait time of 10 minutes seems odd, but not necessarily something to ignore.

There were some suggestions including interference from other magnets and other pieces of kit that are not currently running on the beamline or on other beamlines, the serial cable being quite long can cause issues as well.

Some logging using the genie_python API has been added to emu_routines.py to see if we can catch logs of when the issue occurs again. Asyn trace masks have been applied to DFKPS_01 https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/ASYN-Trace-Masks-(Debugging-IOC,-ASYN)

**Comment by JamesKingWork** [comment]
The timing of the issues last Friday do not correlate to any issues in the log file that we saw, so there's not much to go on there. There is not much more we can do until there is another issue and we can see more logs. I think it's best to put this ticket into awaiting for this reason

**Comment by kjwoodsISIS** [comment]
I disagree with putting this ticket into awaiting.  The Status so Far comment indicates that the Danfysik now reliably reaches the requested field with only short delays - so we have  achieved our acceptance criterion.  I think this ticket can be now be reviewed.

Yes, we saw something odd, which does not correlate to anything we were doing.  But as the above comment says "There is not much more we can do until there is _**another issue**_".  Another issue should be handled via a new ticket.  I think we can create that ticket if the odd thing happens again.

The awaiting label is used for tickets that we cannot start work on, because we are waiting for something else (something definite) to happen first (e.g. for an item of equipment to be made available).  We should not use it to wait for events that might never happen.

**Comment by JamesKingWork** [comment]
I will put it into review in that case

**Comment by JamesKingWork** [comment]
Steve and Koji analysed runs from the start of the cycle and have found so far that they have lost almost 6 hours waiting for danfysik to get to the set field. According to their analysis setting a setpoint on its own is fine, but the issue resides in turning on the danfysik power supply and then setting a setpoint. This has been mostly tested when switching from the zero-field but occurred once when switching from T20 coils when it took 78 minutes.

**Comment by kjwoodsISIS** [comment]
Do we have a copy of Steve & Koji's analysis?  If not, can we get a copy.
   * We should try to correlate with the log files.
   * When was our fix deployed relative to the time period covered by Steve & Koji's analysis?
   * Does it cover the full period, or was our fix deployed part way through?
   * Is there any correlation between the deployment of our fix and the ability to set the field?
   * Have Steve & Koji noticed any difference since we deployed our fix?

**Comment by JamesKingWork** [comment]
The analysis (excel sheet) and email text (txt file) is in shares\ISIS_Experiment_Controls\emu\Danfysik_Diagnosis_20201126...

**Comment by DominicOram** [comment]
> * When was our fix deployed relative to the time period covered by Steve & Koji's analysis?
> * Does it cover the full period, or was our fix deployed part way through?
> * Is there any correlation between the deployment of our fix and the ability to set the field?
> * Have Steve & Koji noticed any difference since we deployed our fix?

It's my understanding that we don't have a fix? We just added extra logging

**Comment by JamesKingWork** [comment]
On combing through log files since we added more logging we have discovered the following pattern:

- Do run with active zero field
- Finish run
- Set genesis and danfysik setpoints to zero
- Set genesys and danfysik power to off
- Set zero field to auto-feedback
- Set zero field from auto-feedback to manual
- Select danfysik as magnet
- Set danfysik power to on
- Attempt to set danfysik field setpoint
- Either
  - cget reveals danfysik power fails to be set to on, retries setting power to on and pass, then can set setpoint correctly (takes a few seconds)
  - cget thinks danfysik power is on, waits for two lots of 5 minutes (first for danfysik to be in setpoint, then to be stable), waitfor exits without having set setpoint correctly, set danfysik power to be on, successfully set setpoint

This seems to me to be linked with setting the danfysik power to be on. We are hoping a short term solution is to when checking that the danfysik power has been set to on, to also check that the setpoint readback value of the field is the expected field (within the tolerance). This however may be covering up an underlying issue with turning on the danfysik, so we should search for a correct solution.

**Comment by JamesKingWork** [comment]
We have not deployed a fix currently

**Comment by JamesKingWork** [comment]
At 23:24:07 and 00:04:05 on 27/11/2020, the danfysik reports a mismatch in protocol and received value on the STATUS. This value eventually gets set to the danfysik POWER record. These timings don't tally with the failures to set setpoints, but it makes me question if there is a communication error it may be because the IOC is reading and setting the POWER record incorrectly.

**Comment by JamesKingWork** [comment]
In meeting with Koji and Steve:

- The delays of ten minutes were timed using the nexus file looking at the timestamps when parameters were first set and when the run was started
- They have analysed past runs from SECI that did the same process of switching from the zero-field to the danfysik and found no such delays or issues turning the danfysik on and getting it to setpoint
- Steve thinks that the zero field is not related i.e. when the zero field is on and you turn the danfysik on via the OPI it works quite happily
- When @davidkeymer has finished looking at the VI to see if it does anything clever with retries etc. I have said we will let them know
- They are going to let us know when they have finished user experiments for the cycle and then we can edit the script to give them a short term fix

Short term fix is to replace lines 220 and 221 from emu_routines.py (the setmag function):
```
while g.cget("danfysik_power")["value"] != "On":
    g.cset("danfysik_power", "On")
```
with:
```
while g.cget("danfysik_power")["value"] != "On" and field - field_range < g.get_value("IN:EMU:DFKPS_01:FIELD:SP:RBV") < field + field_range:
    g.cset("danfysik_power", "On")
    time.sleep(2)
    g.cset("field_Danfysik", field)
```

I have added this apply the fix to the tasks for shutdown.

**Comment by Tom-Willemsen** [comment]
It seems like this ticket hasn't yet met it's acceptance criteria? It's possible I've misunderstood what has been done here...

Status of this ticket as far as I understand it:
> We have not deployed a fix currently

Acceptance criteria of this ticket:
> The Danfysik reliably reaches the requested field with only short delays

It sounds like we have some testing to do when beam is off so I think this ticket is impeded until then?

Additionally, 

> Short term fix is ...

Do we have a ticket for a long term fix? Can you link it to this one please?

**Comment by KathrynBaker** [comment]
Just to clarify, given that this is a timebox, is the rework to create the various follow-up tickets, or to spend more time on this ticket? If there is time left in the timebox more time can be spent, if not this ticket shouldn't be impeded, but there should be new tickets to finish the task.

**Comment by Tom-Willemsen** [comment]
@KathrynBaker I would be fine with either solution, as long as the extra work needed is captured somewhere - either on this ticket by putting it in impeded, or by new ticket(s) specifying what still remains to be done and long-term fix etc.

**Comment by JamesKingWork** [comment]
- Ticket for adding and testing short term fix at https://github.com/ISISComputingGroup/IBEX/issues/5951 and on shutdown task list
- Ticket for investigating long term fix https://github.com/ISISComputingGroup/IBEX/issues/5952
